src/pool/srv_pool.c

```diff
@@ -3628,6 +3765,355 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
+		val = DAOS_RPOP_PO_REDUN_FAC_DEFAULT;
+		rc = rdb_tx_update(tx, &svc->ps_root, &ds_pool_prop_redun_fac, &value);
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade redundancy factor of pool, "
+				"%d.\n", DP_UUID(pool_uuid), rc);
+			D_GOTO(out_free, rc);
+		}
+		need_commit = true;
+	}
+
+	rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_ec_pda, &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_free, rc);
+	} else if (rc == -DER_NONEXIST) {
+		val = DAOS_PROP_PO_EC_PDA_DEFAULT;
+		rc = rdb_tx_update(tx, &svc->ps_root, &ds_pool_prop_ec_pda, &value);
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade EC performance domain "
+				"affinity of pool, %d.\n", DP_UUID(pool_uuid), rc);
+			D_GOTO(out_free, rc);
+		}
+		need_commit = true;
+	}
+
+	rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_rp_pda, &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_free, rc);
+	} else if (rc == -DER_NONEXIST) {
+		val = DAOS_PROP_PO_RP_PDA_DEFAULT;
+		rc = rdb_tx_update(tx, &svc->ps_root, &ds_pool_prop_rp_pda, &value);
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade RP performance domain "
+				"affinity of pool, %d.\n", DP_UUID(pool_uuid), rc);
+			D_GOTO(out_free, rc);
+		}
+		need_commit = true;
+	}
+
+	d_iov_set(&value, &val32, sizeof(val32));
+	rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_upgrade_status, &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_free, rc);
+	} else if (rc == -DER_NONEXIST || val32 != DAOS_UPGRADE_STATUS_IN_PROGRESS) {
+		val32 = DAOS_UPGRADE_STATUS_IN_PROGRESS;
+		rc = rdb_tx_update(tx, &svc->ps_root, &ds_pool_prop_upgrade_status, &value);
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade 'upgrade status' "
+				"of pool, %d.\n", DP_UUID(pool_uuid), rc);
+			D_GOTO(out_free, rc);
+		}
+		need_commit = true;
+	}
+
+	rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_upgrade_global_version,
+			   &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_free, rc);
+	} else if (rc == -DER_NONEXIST || val32 != DS_POOL_GLOBAL_VERSION) {
+		val32 = DS_POOL_GLOBAL_VERSION;
+		rc = rdb_tx_update(tx, &svc->ps_root,
+				   &ds_pool_prop_upgrade_global_version, &value);
+		if (rc != 0) {
+			D_ERROR("failed to write upgrade global version prop, "DF_RC"\n",
+				DP_RC(rc));
+			D_GOTO(out_free, rc);
+		}
+		need_commit = true;
+	}
+
+	if (need_commit) {
+		daos_prop_t *prop = NULL;
+
+		rc = rdb_tx_commit(tx);
+		if (rc)
+			D_GOTO(out_free, rc);
+		rc = pool_prop_read(tx, svc, DAOS_PO_QUERY_PROP_ALL, &prop);
+		if (rc)
+			D_GOTO(out_free, rc);
+		rc = ds_pool_iv_prop_update(svc->ps_pool, prop);
+	}
+
+out_free:
+	D_FREE(hdl_uuids);
+	return rc;
+}
+
+static int ds_pool_mark_upgrade_completed(uuid_t pool_uuid,
+					  struct pool_svc *svc, int rc)
+{
+	struct rdb_tx			tx;
+	d_iov_t				value;
+	uint32_t			val;
+	int				rc1;
+	bool				need_put_leader = false;
+	daos_prop_t			*prop = NULL;
+
+	if (!svc) {
+		rc1 = pool_svc_lookup_leader(pool_uuid, &svc, NULL);
+		if (rc1 != 0)
+			return rc1;
```

Comment:
Do we need

    need_put_leader = true;

Reply:
Yes, looks we don't need such logic, caller always pass svc.